packages/google_sign_in/lib/google_sign_in.dart

```diff
@@ -138,18 +134,42 @@ class GoogleSignIn {
   GoogleSignInAccount _currentUser;
   GoogleSignInAccount get currentUser => _currentUser;
 
+  /// Completers used for memoization of signIn and signOut processes.
+  Completer<GoogleSignInAccount> _signInCompleter;
+  Completer<GoogleSignInAccount> _signOutCompleter;
+
   /// Attempts to sign in a previously authenticated user without interaction.
-  Future<GoogleSignInAccount> signInSilently() => _callMethod('signInSilently');
+  Future<GoogleSignInAccount> signInSilently() =>
+      _signInWithMethod('signInSilently');
 
   /// Starts the sign-in process.
-  Future<GoogleSignInAccount> signIn() => _callMethod('signIn');
+  Future<GoogleSignInAccount> signIn() => _signInWithMethod('signIn');
+
+  Future<GoogleSignInAccount> _signInWithMethod(String method) {
+    if (_currentUser != null) return new Future.value(_currentUser);
+    if (_signInCompleter != null) return _signInCompleter.future;
+    _signInCompleter = new Completer();
+    _signInCompleter.complete(_callMethod(method).whenComplete(() {
+      _signInCompleter = null;
+    }));
+    return _signInCompleter.future;
+  }
 
   /// Marks current user as being in the signed out state.
-  Future<GoogleSignInAccount> signOut() => _callMethod('signOut');
+  Future<GoogleSignInAccount> signOut() => _signOutWithMethod('signOut');
 
   /// Disconnects the current user from the app and revokes previous
   /// authentication.
-  Future<GoogleSignInAccount> disconnect() => _callMethod('disconnect');
+  Future<GoogleSignInAccount> disconnect() => _signOutWithMethod('disconnect');
+
+  Future<GoogleSignInAccount> _signOutWithMethod(String method) {
+    if (_signOutCompleter != null) return _signOutCompleter.future;
+    _signOutCompleter = new Completer();
+    _signOutCompleter.complete(_callMethod(method).whenComplete(() {
```

Comment:
Instead of introducing a completer, can you not just hold on to the Future returned by _callMethod?